src/pypwsqc/bias_correction.py

```diff
@@ -0,0 +1,198 @@
+    raindata: np.array
+        Data larger than zero
+    raindata_trs: np.array
+        Data larger than threshold
+    p0: float
+        Probability of no rainfall
+    """
+    data_valid = data[~np.isnan(data)]
+    raindata = data_valid[data_valid > 0]
+    raindata_trs = raindata[raindata > threshold]
+
+    # First quick check for sufficient data length
+    if data_valid.shape[0] < 10 or raindata_trs.shape[0] < 5:
+        return np.nan, np.nan, np.nan, np.nan
+
+    p0 = np.mean(data_valid <= 0)
+
+    # statistics for initial guesses
+    raindata_mean = np.mean(raindata)
+    raindata_variance = np.var(raindata)
+
+    if raindata_variance == 0:
+        msg = "Invalid data: no variance in your data!"
+        raise ValueError(msg)
+
+    # initial guess for parameters (Method of Moments)
+    b_init = raindata_variance / raindata_mean
+    a_init = raindata_mean / b_init
+    initial_guess = [a_init, b_init]
+
+    return initial_guess, raindata, raindata_trs, p0
+
+
+def _negative_log_likelihood(initial_guess, threshold, raindata, raindata_trs):
+    """
+    Define the log-likelihood function.
+
+     Define the log-likelihood function which will be optimized in the
+     fit_gamma_params_with_threshold function.
+
+    Parameters
+    ----------
+    initial_guess: [a_init, b_init]
+        Method of Moments guess for shape and scale parameters of the gamma distribution
+    threshold : float
+        Threshold for censoring data
+    raindata: np.array
+        Data larger than zero
+    raindata_trs: np.array
+        Data larger than threshold
+
+    Returns
+    -------
+    float:
+        The negative log-likelihood value for initial guess
+
+    """
+    aa, bb = initial_guess
+
+    if aa == 0 or bb == 0:
+        msg = "Gamma Dist not defined for params == 0!"
+        raise ValueError(msg)
+
+    y = stats.gamma.pdf(raindata_trs, a=aa, loc=0, scale=bb)
+    pa = stats.gamma.cdf(threshold, a=aa, loc=0, scale=bb)
+
+    ltest = np.sum(np.log(y)) + np.log(pa) * (raindata.shape[0] - raindata_trs.shape[0])
+    return -ltest
+
+
+def fit_gamma_params_with_threshold(
+    data,
+    threshold,
+) -> tuple[float, float, float]:
+    """Fit a censored gamma distribution to rainfall data.
+
+    Parameters
+    ----------
+    data : np.array
+        Data for which the theoretical gamma distribution is fitted
+    threshold : float
+        Threshold for censoring data (detection threshold)
+
+    Returns
+    -------
+    a_opt: float
+        shape parameter for the gamma distribution
+    b_opt: float
+        scale parameter for the gamma distribution
+    p0: float
+        Probability of no rainfall
+    """
+    try:
+        initial_guess, raindata, raindata_trs, p0 = _data_preprocessing(data, threshold)
+    except ValueError:
+        print("Preprocessing failed")
+        return np.nan, np.nan, np.nan
+
+    # Handle failed preprocessing

```

Comment:
Why do we need to handle failed preprocessing if we return nan above in this case anyway?